subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/JavaLanguageVersion.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain;
+
+import org.gradle.api.Incubating;
+import org.gradle.jvm.toolchain.internal.DefaultJavaLanguageVersion;
+
+/**
+ * Represents a Java Language version
+ *
+ * @since 6.7
+ */
+@Incubating
+public interface JavaLanguageVersion extends Comparable<JavaLanguageVersion> {
+
+    static JavaLanguageVersion of(int version) {
+        return new DefaultJavaLanguageVersion(version);
```

Comment:
:o: I would special case a number of versions, in order to reduce memory usage (a bit like the small numbers in `Integer`).